src/librustc/middle/traits/mod.rs

```diff
@@ -95,15 +96,27 @@ pub struct ObligationCause<'tcx> {
     pub code: ObligationCauseCode<'tcx>
 }
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq)]
 pub enum ObligationCauseCode<'tcx> {
     /// Not well classified or should be obvious from span.
     MiscObligation,
 
+    /// Obligation that triggers warning until RFC 1214 is fully in place.
+    RFC1214(Rc<ObligationCauseCode<'tcx>>),
```

Comment:
@arielb1 

> How do you ensure these don't pollute any cache? I mean, FulfillmentContext::duplicate_set, ty::ctxt::fulfilled_predicates?

So, I thought about this, but I wasn't too worried because I figured if you are getting these RFC1214 warnings, it's (morally) as good as an error. Still, we don't want to ICE etc, so I suppose you are right that I ought to patch up the cache behavior.

There are two dangerous scenarios:

- fulfillment context's `duplicate_set` always contains all predicates that we *attempted* to solve, whether or not it has ended in error. The danger then is that we add an RFC1214 warning to the cache, report a warning, and then later have the same case occur, but where an error would be triggered. This would not lead to us accepting programs we shouldn't, but I imagine it could lead to an ICE. For example, one might trigger an `T:Eq` obligation from a type like `fn(MustBeEq<T>)`, which will yield a warning, and then in the same function write other code that requires `T:Eq`. That other code should error, but would not, and if we get to trans we could have a failure. Not sure how likely that case is, but I guess it wouldn't be too hard to sidestep it by tweaking the cache strategy.

- A similar scenario can play out across functions via the `fulfilled_predicates` cache in the tcx.

For the latter, I imagine the case is to check the origin of the predicate and take that into account when we consider whether a failure of this obligation will lead to an error. For the former, probably the easiest thing to do is to add a boolean to the key ("is warning" vs "is error"). Hacky, but temporary.

Make sense?